### ISSUE TITLE
[contracts] use `solc` for 10x faster contract compilation on CI

### ISSUE THREAD

**Issue by snario** [counterfactual/monorepo#393]
When #392 is done we can use `ethereum-waffle`'s latest version to use `solc` which speeds up contracts compilation 10x

**Comment by snario** [comment]
Should be straightforward to install `solc` into CircleCI's environment to do this.

**Comment by AlexXiong97** [comment]
Ahh... sorry for spamming this. CI keeps failing (as the docker image uses Ubuntu14 which doesn't have some dependency packages for solc and make things a bit messy, trying to use another image)

I will commit amend to another name to avoid excessive referencing

**Comment by AlexXiong97** [comment]
**TL;DR**: downloading `solc` in CircleCI is bumpy (hope I'm wrong), and installation time exceeds time 
saved from fast compilation, thus almost not worth it to change.

Time saved: when run locally, compilation of the `contracts` package reduced from 9s to 2s.

It turns out that getting ci to work is not that trivial. 

I first tried using PPA which returns me `solc: Depends on libcvc4-5 but not installable` error which indeed is not a package and there's no online answer related to this. 

Then I tried to use `snap install` to workaround, which turns out also not possible in CI docker, due to the restriction on [`systemd` and `systemctl`](https://stackoverflow.com/questions/39169403/systemd-and-systemctl-within-ubuntu-docker-images), two of which are needed to get `snapd` to work.

 Then I was forced to try PPA again, and two new problems emerge:
1.  `software-properties-common` package is missing to enable `add-apt-repository` in Debian
2. `keyring` package is also missing and most `gpg keyserver` are either unreachable or request timeout, and ultimately `gpg --keyserver ha.pool.sks-keyservers.net --recv-keys 1C52189C923F6CA9` is able to retrieve the public key for the solidity release.

And then.. the weird `libcvc4-5` problem occurs again. 

Now, I'm trying to build from source directly, which requires me to install `CMake` and `Boost` etc.. which makes the time saved from fast compilation not worth it. 

[my embarrassing struggle to get it work](https://github.com/AlexXiong97/monorepo/commits/alex/solc)

**Comment by snario** [comment]
> and installation time exceeds time saved from fast compilation, thus almost not worth it to change.

Surely we would cache the installation?

**Comment by AlexXiong97** [comment]
ah, sorry. You're right. I'll try to get it done.

**Comment by AlexXiong97** [comment]
I give another "still failing" attempt, [see here](https://github.com/AlexXiong97/monorepo/commit/2b26bb68f94f8cd3545d68f7ae0241c37e489127#diff-1d37e48f9ceff6d8030570cd36286a61R34). 


To allow installing from PPA, I set `machine=true` instead of using the node10 orbs,
which allow me to successfully install `solc`'s binary to `/usr/bin/solc`. 

I previously tried to cache it, and restore it, however `permission denied` error would occur. 
I once tried to `sudo chmod +R 777 /usr/bin/` before storing cache, which indeed allow me to run restore `solc` and not have to download everytime, 
however such approach will result in changing `/usr/bin/sudo` program as well, which will cause other problem further down the road. 

In CircleCI's discuss forum, it is recommended not to cache those from the system root level, and download them per job instead. 

Later on I manage not to solve all problems above, the compiler version of `solc` downloaded is the latest (i.e. `0.5.7`) whereas the contracts are now using `0.4.6`, thus still failing. 

![Screenshot from 2019-04-01 00-29-54](https://user-images.githubusercontent.com/20514086/55291766-4d556800-5415-11e9-9fa9-54456d7268c2.png)

**Comment by snario** [comment]
@AlexXiong97 can you make this a branch on this repo please?

**Comment by AlexXiong97** [comment]
@snario but currently I don't think I have the permission to though, only team member/collaborator can push branch to this?

**Comment by AlexXiong97** [comment]
I'm trying to customize a Docker image to reduce all the pain now and make a cleaner circleci config file, as recommended for version 2 .

**Comment by AlexXiong97** [comment]
Update:
1. I publish a customized docker image specifically for the build job. Instead of using `circleci/node10`, I used a Ubuntu with node10 and yarn 1.15 and solc binary installed. 

2. update the `waffle.json` from loading solc.js to `native`

3. beautify the original `config.yml` a bit

Result:
Even the run time vary every ci run, the `build` job is indeed faster, **down to `2:42` from `3:29`** of the current master branch. 
Time spending on the rest is roughly the same. 

TODO: (maybe another issue?)
we could further cut down the `build` time if we pre-load the customized image without pulling every time.
i.e. getting rid of : 
```log
Starting container web3docker/counterfactual:0.0.1
  image cache not found on this host, downloading web3docker/counterfactual:0.0.1
0.0.1: Pulling from web3docker/counterfactual
```
